penn/studyspaces.py

```diff
@@ -28,18 +28,101 @@ def get_buildings(self):
 
         soup = BeautifulSoup(requests.get("{}/spaces".format(BASE_URL)).content, "html5lib")
         options = soup.find("select", {"id": "lid"}).find_all("option")
-        return [{"id": int(opt["value"]), "name": str(opt.text), "service": "libcal"} for opt in options]
+        return [{"id": int(opt["value"]), "name": str(opt.text), "service": "libcal"} for opt in options if int(opt["value"]) > 0]
+
+    def book_room(self, building, room, start, end, firstname, lastname, email, groupname, phone, size, fake=False):
+        """Books a room given the required information.
+
+        :param building:
+            The ID of the building the room is in.
+        :type building: int
+        :param room:
+            The ID of the room to book.
+        :type room: int
+        :param start:
+            The start time range of when to book the room.
+        :type start: datetime
+        :param end:
+            The end time range of when to book the room.
+        :type end: datetime
+        :param fake:
+            If this is set to true, don't actually book the room. Default is false.
+        :type fake: bool
+        :returns:
+            Boolean indicating whether the booking succeeded or not.
+        :raises ValueError:
+            If one of the fields is missing or incorrectly formatted, or if the server fails to book the room.
+        """
+
+        data = {
+            "formData[fname]": firstname,
+            "formData[lname]": lastname,
+            "formData[email]": email,
+            "formData[nick]": groupname,
+            "formData[q2533]": phone,
+            "formData[q2555]": size,
+            "forcedEmail": ""
+        }
+
+        try:
+            room_obj = self.get_room_id_name_mapping(building)[room]
+        except KeyError:
+            raise ValueError("No room with id {} found in building with id {}!".format(room, building))
+
+        if room_obj["lid"] != building:
+            raise ValueError("Mismatch between building IDs! (expected {}, got {})".format(building, room_obj["lid"]))
+
+        room_data = {
+            "id": 1,
+            "eid": room,
+            "gid": room_obj["gid"],
+            "lid": room_obj["lid"],
+            "start": start.strftime("%Y-%m-%d %H:%M"),
+            "end": end.strftime("%Y-%m-%d %H:%M")
+        }
+
+        for key, val in room_data.items():
+            data["bookings[0][{}]".format(key)] = val
+
+        if fake:
+            return True
```

Comment:
?

Reply:
I added that argument for testing, just to make sure that all of the steps up to the final booking step worked correctly.